File: Cpp/Search a 2D Matrix.cpp
class Solution {
public:
    /**
     * @param matrix, a list of lists of integers
     * @param target, an integer
     * @return a boolean, indicate whether matrix contains target
     */
    bool searchMatrix(vector<vector<int> > &matrix, int target) {
        // write your code here
        if (matrix.empty() || matrix[0].empty()) return false;
        int m = matrix.size(), n = matrix[0].size();
        int start = 0, end = m * n - 1, mid;
        while (start + 1 < end) {
            mid = start + (end - start) / 2;
            int cur = matrix[mid / n][mid % n];
            if (cur == target) return true;
            else if (cur > target) end = mid;
            else start = mid;
        }
        if (matrix[start / n][start % n] == target) return true;
        else if (matrix[end / n][end % n] == target) return true;
        return false;
    }
};

File: Python/Search-a-2D-Matrix.py
class Solution:
    """
    @param matrix, a list of lists of integers
    @param target, an integer
    @return a boolean, indicate whether matrix contains target
    """
    def searchMatrix(self, matrix, target):
        # write your code here
        lenx = len(matrix)
        if lenx == 0 or matrix is None:
            return False
        leny = len(matrix[0])
        left = 0
        right = lenx * leny - 1
        while left <= right:
            mid = left + (right - left) / 2
            x = mid / leny
            y = mid % leny
            if matrix[x][y] == target:
                return True
            elif matrix[x][y] < target:
                left = mid + 1
            else:
                right = mid - 1
        return False
